swarm/pss/protocol.go

```diff
@@ -0,0 +1,201 @@
+package pss
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/protocols"
+	"github.com/ethereum/go-ethereum/rlp"
+	whisper "github.com/ethereum/go-ethereum/whisper/whisperv5"
+	"time"
+)
+
+// protocol specification of the pss capsule
+var pssSpec = &protocols.Spec{
+	Name:       "pss",
+	Version:    1,
+	MaxMsgSize: defaultMaxMsgSize,
+	Messages: []interface{}{
+		PssMsg{},
+	},
+}
+
+// Bridges pss send/receive with devp2p protocol send/receive
+//
+// Implements p2p.MsgReadWriter
+type PssReadWriter struct {
+	*Pss
+	LastActive time.Time
+	rw         chan p2p.Msg
+	spec       *protocols.Spec
+	topic      *whisper.TopicType
+	sendFunc   func(string, whisper.TopicType, []byte) error
+	key        string
+}
+
+// Implements p2p.MsgReader
+func (prw PssReadWriter) ReadMsg() (p2p.Msg, error) {
+	msg := <-prw.rw
+	log.Trace(fmt.Sprintf("pssrw readmsg: %v", msg))
+	return msg, nil
+}
+
+// Implements p2p.MsgWriter
+func (prw PssReadWriter) WriteMsg(msg p2p.Msg) error {
+	log.Trace("pssrw writemsg", "msg", msg)
+	rlpdata := make([]byte, msg.Size)
+	msg.Payload.Read(rlpdata)
+	pmsg, err := rlp.EncodeToBytes(ProtocolMsg{
+		Code:    msg.Code,
+		Size:    msg.Size,
+		Payload: rlpdata,
+	})
+	if err != nil {
+		return err
+	}
+	return prw.sendFunc(prw.key, *prw.topic, pmsg)
+}
+
+// Injects a p2p.Msg into the MsgReadWriter, so that it appears on the associated p2p.MsgReader
+func (prw PssReadWriter) injectMsg(msg p2p.Msg) error {
+	log.Trace(fmt.Sprintf("pssrw injectmsg: %v", msg))
+	prw.rw <- msg
+	return nil
+}
+
+// For devp2p protocol integration only.
+//
+// Convenience object for passing messages in and out of the p2p layer
+type PssProtocol struct {
+	*Pss
+	proto        *p2p.Protocol
+	topic        *whisper.TopicType
+	spec         *protocols.Spec
+	pubKeyRWPool map[string]p2p.MsgReadWriter
+	symKeyRWPool map[string]p2p.MsgReadWriter
+	flags        byte
+}
+
+// For devp2p protocol integration only.
+//
+// Maps a Topic to a devp2p protocol.
+//
+// flags: 0x01 = asymmetric messaging enabled, 0x02 symmetric messaging enabled
```

Comment:
I heavily suggest making `flags` an explicit "options" type for readability, for example:

```go
type EncryptionOpts struct {
	Symmetric  bool
	Asymmetric bool
}
```

Reply:
Confound it, I thought I'd found a way to sneak in some C-style goodies.